Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp

```diff
@@ -292,14 +292,14 @@ void RemoteRenderingBackendProxy::releaseRemoteImageBufferSet(RemoteImageBufferS
     send(Messages::RemoteRenderingBackend::ReleaseRemoteImageBufferSet(bufferSet.identifier()));
 }
 
-void RemoteRenderingBackendProxy::moveToSerializedBuffer(WebCore::RenderingResourceIdentifier identifier)
+void RemoteRenderingBackendProxy::moveToSerializedBuffer(RemoteSerializedImageBufferReference ref)
 {
-    send(Messages::RemoteRenderingBackend::MoveToSerializedBuffer(identifier));
+    send(Messages::RemoteRenderingBackend::MoveToSerializedBuffer(ref));
```

Comment:
With the types, the proposed code here says:
- Destroy SerializedImageBuffer with name `ref`
- Create new SerializedImageBuffer ...

So you can see that somehow the argument of the operation is already something of wrong type. The operation targets a ImageBuffer, but it gets a  SerializedImageBuffer ref, which cannot possibly exist at the time of the call. So the code doesn't make that much sense.

Reply:
My understanding is that removing slowness in `RemoteImageBufferProxy::sinkIntoSerializedImageBuffer` uncovered a race.
The race happens when an `ImageBuffer` is moved from GPUP thread A to thread B and then back to thread A.
If thread B is slow, then thread A will add the `ImageBuffer` to `RemoteRenderingBackend::m_sharedResourceCache` and then will itself take it.
Thread A has no way of knowing it should wait for thread B to take the `ImageBuffer` and then put it back in `RemoteRenderingBackend::m_sharedResourceCache`.
The result is thread B times out while trying to take the `ImageBuffer` (because thread A already took it), causing a crash (`ASSERT_IS_TESTING_IPC` in `RemoteRenderingBackend::moveToImageBuffer`).

A different solution would be to give the `ImageBuffer` a new name (`WebCore::RenderingResourceIdentifier`) when it moves to a different GPUP thread.
I decided against this because I thought we may have messages about the `ImageBuffer` in flight as it moves or we may want to support that in the future.